File: src/correct_controller/src/twist_relay.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <geometry_msgs/msg/twist_stamped.hpp>

using  std::placeholders::_1;


class TwistRelay : public rclcpp::Node
{
    public: TwistRelay() : Node("twist_relay")
    {
        controller_sub = create_subscription<geometry_msgs::msg::Twist>("/correct_controller/cmd_vel_unstamped",10,std::bind(&TwistRelay::controller_twist_callback, this, _1));
        controller_pub = create_publisher<geometry_msgs::msg::TwistStamped>("/correct_controller/cmd_vel",10);
        joy_sub = create_subscription<geometry_msgs::msg::TwistStamped>("/input_joy/cmd_vel_stamped",10,std::bind(&TwistRelay::joy_twist_callback,this,_1));
        joy_pub = create_publisher<geometry_msgs::msg::Twist>("/input_joy/cmd_vel",10);

    }

    private:
        rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr controller_sub;
        rclcpp::Subscription<geometry_msgs::msg::TwistStamped>::SharedPtr joy_sub;
        rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr controller_pub;
        rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr joy_pub;

        geometry_msgs::msg::Twist twist;
        geometry_msgs::msg::TwistStamped twist_stamped;


        void controller_twist_callback(const geometry_msgs::msg::Twist &msg)
        {
            twist_stamped.header.stamp = get_clock()->now();
            twist_stamped.twist = msg;
            controller_pub->publish(twist_stamped);
        }

        void joy_twist_callback(const geometry_msgs::msg::Twist &msg)
        {
            twist = msg;
            joy_pub->publish(twist);
        }
    };


int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TwistRelay>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: install/correct_controller/local/lib/python3.10/dist-packages/correct_controller/twist_relay.py
#!/usr/bin/env python3
# Subscribe from the twist mux and publish unto correct_controller/cmd_vel in order to move the robot
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist, TwistStamped


class TwistRelay(Node):
    def __init__(self):
        super().__init__("twist_relay")
        self.controller_sub = self.create_subscription(Twist, "/correct_controller/cmd_vel_unstamped", self.controller_twist_callback, 10)
        self.controller_pub = self.create_publisher(TwistStamped, "/correct_controller/cmd_vel",10)
        self.joy_sub = self.create_subscription(TwistStamped, "/input_joy/cmd_vel_stamped",self.joy_twist_callback,10)
        self.joy_pub = self.create_publisher(Twist, "/input_joy/cmd_vel",10)

    #Add the Stamped to the Twist message from twist_mux
    def controller_twist_callback(self, msg):
        twist_stamped = TwistStamped()
        twist_stamped.header.stamp = self.get_clock().now().to_msg()
        twist_stamped.twist = msg
        self.controller_pub.publish(twist_stamped)

    
    #Convert the twist_stamped velocities on the joystick to twist
    def joy_twist_callback(self, msg):
        twist = Twist()
        twist = msg.twist
        self.joy_pub.publish(twist)


def main():
    rclpy.init()
    node = TwistRelay()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

